### PR TITLE
HWSTL-35 Create a C++ implementation of Myers's diff algorithm

### PR DESCRIPTION
**Pull request by JulianvDoorn**

### CODE REVIEW

File: include/diff.hpp [hunk 1]
```diff
@@ -9,28 +9,86 @@
...
+ * Current implementation only works for linear storage containers, like
+ * vectors and strings. sequence_view is not able to mutate the viewed type.
+ * 
+ * Future specializations for mutable sequences and non-linear stored sequences
+ * are possible but currently not supported.
+ * 
+ * @tparam t_sequenced_type Type in sequence viewed as a sequence_view
+ */
+template <class t_sequenced_type>
 class sequence_view {
```
Comment (JulianvDoorn) [inline include/diff.hpp]: Class should be moved into its own file

File: include/diff.hpp [hunk 2]
```diff
@@ -89,27 +190,29 @@ enum class PatchOperation {
 class Patch {
```
Comment (JulianvDoorn) [inline include/diff.hpp]: Inconsistent class naming

File: include/diff.hpp [hunk 3]
```diff
@@ -130,165 +233,188 @@ class Patch {
...
+         * Based off of Robert Elder's optimized Myer's diff
+         * See more: http://blog.robertelder.org/diff-algorithm/
+         * 
+         * @tparam t_sequenced_type Type contained in the sequence
+         * @param lhs_seq Source sequence
+         * @param rhs_seq Destination sequence to generate patches to
+         * @param patches Patches holding vector
+         */
+        template <class t_sequenced_type>
+        void _diff(const sequence_view<t_sequenced_type>& lhs_seq, const sequence_view<t_sequenced_type>& rhs_seq, std::vector<Patch>& patches) {
```
Comment (JulianvDoorn) [inline include/diff.hpp]: Really long function, could be split into two

File: include/diff.hpp (hunk 2 above)
Comment (JulianvDoorn) [reply include/diff.hpp]: A company-wide naming scheme should be setup

**Comment by BountyVSGames** [comment]
You're sometimes not really consistent with where you end you scope. Like in this example you first close the previous if statement, and then you break off the line before the else if. While in the other cases you don't do that. It's not that bad if your the only person working on the project but if you have more then 1 programmer working on it, it could really piss people off.
image

For the rest I don't really have anything to say about it, since I'm not that familiar with C++ yet.

File: include/diff.hpp [hunk 4]
```diff
@@ -1,147 +1,21 @@
...
-        int32_t x_values_len = 2 * std::min(lhs_size, rhs_size) + 2;
+namespace diff {
+    template <class t_sequenced_type>
+    class _diff {
+        const sequence_view<t_sequenced_type>& lhs_seq;
+        const sequence_view<t_sequenced_type>& rhs_seq;
+        std::vector<patch>& patches;
+        int32_t lhs_size;
+        int32_t rhs_size;
+        int32_t max_len;
```
Comment (wvanbreukelen) [inline include/diff.hpp:17]: Length should always be unsigned, `uint32_t` is the better option here. Even better is to use `size_t`

File: include/patch.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * Patches are differences between sequences. If there is a patch between
+ * sequence `A` and sequence `B`, applying a patch onto sequence `A` will,
+ * possibly partially, turn sequence `A` into `B`
+ */
+class patch {
+    patch_operation m_op;
+    size_t m_begin;
+    std::string m_seq;
+
+    patch(patch_operation op, size_t begin, std::string&& str) :
```
Comment (wvanbreukelen) [inline include/patch.hpp:24]: patch_operation and begin could be declared const

File: include/patch.hpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        m_op(op),
+        m_begin(begin),
+        m_seq(str)
+    { }
+
+    patch(patch_operation op, size_t begin, const std::string& str) :
```
Comment (wvanbreukelen) [inline include/patch.hpp:30]: Same here

File: include/sequence_view.hpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+     * @brief Prints the sequence_view to a t_os
+     * 
+     * @tparam t_os ostream type
+     * @param os ostream
+     * @param seq sequence_view
+     * @return t_os& Chainable ostream
+     */
+    template <class t_os>
+    friend t_os& operator<< (t_os& os, const sequence_view& seq) {
+        if (seq.m_seq != nullptr) {
```
Comment (wvanbreukelen) [inline include/sequence_view.hpp:164]: I would write:

```cpp
if (seq.m_seq) {
  // Code...
} else {
  // os << "Null sequence";
}

File: src/main.cpp [hunk 8]
```diff
@@ -1,22 +1,24 @@
...
-    Diff a;
-
     auto lhs = std::string("goore wormen");
     auto rhs = std::string("guur feesten");
 
     std::cout << lhs << std::endl;
     std::cout << rhs << std::endl;
 
-    auto v = a.diff(lhs, rhs);
+    auto v = diff::diff(lhs, rhs);
```
Comment (wvanbreukelen) [inline src/main.cpp:12]: Weird namespacing, I would rename it to ```diff::perform(lhs, rhs)```

File: src/main.cpp [hunk 9]
```diff
@@ -1,22 +1,24 @@
...
 
     for (auto p : v) {
         std::cout << p << std::endl;
     }
 
-    system("pause");
+    // system("pause");
```
Comment (wvanbreukelen) [inline src/main.cpp:18]: Nitpicking: could be removed

File: src/main.cpp [hunk 10]
```diff
@@ -1,22 +1,24 @@
...
+
+    while (true) { }
+
+    int i = 10;
```
Comment (wvanbreukelen) [inline src/main.cpp:22]: Remove it

**Review by wvanbreukelen — CHANGES_REQUESTED** [review]
Some small changes preferred.

File: src/main.cpp (hunk 9 above)
Comment (JulianvDoorn) [reply src/main.cpp:18]: Should be removed*

File: src/main.cpp (hunk 10 above)
Comment (JulianvDoorn) [reply src/main.cpp:22]: Cleanup the main()

File: src/main.cpp (hunk 8 above)
Comment (JulianvDoorn) [reply src/main.cpp:12]: By writing diff::perform, the descriptive naming of the function "diff" is completely lost. What if someone performs `using namespace diff;`? Then obscure function calls to `perform(lhs, rhs)` will be all the way over...

**Pull request merged**